packages/next/next-server/server/image-optimizer.ts

```diff
@@ -0,0 +1,230 @@
+import { parse } from 'url'
+import { IncomingMessage, ServerResponse } from 'http'
+import { join } from 'path'
+import { mediaType } from '@hapi/accept'
+import { createReadStream, promises } from 'fs'
+import { createHash } from 'crypto'
+import Server from './next-server'
+import { fileExists } from '../../lib/file-exists'
+
+let sharp: typeof import('sharp')
+//const AVIF = 'image/avif'
+const WEBP = 'image/webp'
+const PNG = 'image/png'
+const JPEG = 'image/jpeg'
+const MIME_TYPES = [/* AVIF, */ WEBP, PNG, JPEG]
+const CACHE_VERSION = 1
+
+export async function imageOptimizer(
+  server: Server,
+  req: IncomingMessage,
+  res: ServerResponse
+) {
+  const { nextConfig, distDir } = server
+  const { sizes = [], domains = [] } = nextConfig?.experimental?.images || {}
+  const { url: reqUrl = '/', headers } = req
+  const { query } = parse(reqUrl, true)
+  const { url, w, q } = query
+  const proto = headers['x-forwarded-proto'] || 'http'
+  const host = headers['x-forwarded-host'] || headers.host
+  const mimeType = mediaType(req.headers.accept, MIME_TYPES) || ''
+
+  if (!url) {
+    res.statusCode = 400
+    res.end('"url" parameter is required')
+    return { finished: true }
+  } else if (Array.isArray(url)) {
+    res.statusCode = 400
+    res.end('"url" parameter cannot be an array')
+    return { finished: true }
+  }
+
+  let absoluteUrl: URL
+  try {
+    absoluteUrl = new URL(url)
+
+    if (
+      Array.isArray(domains) &&
+      domains.length > 0 &&
+      !domains.includes(absoluteUrl.hostname)
```

Comment:
Does this mean `domains` is optional?

Reply:
Looks like we should validate `protocol`

Reply:
@nkzawa The `domains` property is optional if the user only intends to use relative URLs. If they use absolute URLs (aka external) then they must define `domains`.

Also, I added validation for `protocol`, thanks!